src/aiori-CEPHFS.c

```diff
@@ -0,0 +1,361 @@
+        int ret;
+
+        if (access == WRITE)
+        {
+                ret = ceph_write(cmount, fd, buf, size, param->offset);
+                if (ret < 0) {
+                        CEPHFS_ERR("unable to write file to CephFS", ret);
+                } else if (ret < size) {
+                        CEPHFS_ERR("short write to CephFS", ret);
+                }
+                if (param->fsyncPerWrite == TRUE) {
+                        CEPHFS_Fsync(&fd, param);
+                }
+        }
+        else /* READ */
+        {
+                ret = ceph_read(cmount, fd, buf, size, param->offset);
+                if (ret < 0) {
+                        CEPHFS_ERR("unable to read file from CephFS", ret);
+                } else if (ret < size) {
+                        CEPHFS_ERR("short read from CephFS", ret);
+                }
+
+        }
+        return length;
+}
+
+static void CEPHFS_Fsync(void *file, IOR_param_t * param)
+{
+        int fd = *(int *) file;
+        int ret = ceph_fsync(cmount, fd, 0);
+        if (ret < 0) {
+                CEPHFS_ERR("ceph_fsync failed", ret);
+        }
+}
+
+static void CEPHFS_Close(void *file, IOR_param_t * param)
+{
+        int fd = *(int *) file;
+        int ret = ceph_close(cmount, fd);
+        if (ret < 0) {
+                CEPHFS_ERR("ceph_close failed", ret);
+        }
+        free(file);
+
+        return;
+}
+
+static void CEPHFS_Delete(char *testFileName, IOR_param_t * param)
+{
+        int ret = ceph_unlink(cmount, pfix(testFileName));
+        if (ret < 0) {
+                CEPHFS_ERR("ceph_unlink failed", ret);
+        }
+        return;
+}
+
+static IOR_offset_t CEPHFS_GetFileSize(IOR_param_t * param, MPI_Comm testComm,
+                                      char *testFileName)
+{
+        struct stat stat_buf;
+        IOR_offset_t aggFileSizeFromStat, tmpMin, tmpMax, tmpSum;
+
+        int ret = ceph_stat(cmount, pfix(testFileName), &stat_buf);
+        if (ret < 0) {
+                CEPHFS_ERR("ceph_stat failed", ret);
+        }
+        aggFileSizeFromStat = stat_buf.st_size;
+
+        if (param->filePerProc == TRUE) {
+                MPI_CHECK(MPI_Allreduce(&aggFileSizeFromStat, &tmpSum, 1,
+                                        MPI_LONG_LONG_INT, MPI_SUM, testComm),
+                          "cannot total data moved");
+                aggFileSizeFromStat = tmpSum;
+        } else {
+                MPI_CHECK(MPI_Allreduce(&aggFileSizeFromStat, &tmpMin, 1,
+                                        MPI_LONG_LONG_INT, MPI_MIN, testComm),
+                          "cannot total data moved");
+                MPI_CHECK(MPI_Allreduce(&aggFileSizeFromStat, &tmpMax, 1,
+                                        MPI_LONG_LONG_INT, MPI_MAX, testComm),
+                          "cannot total data moved");
+                if (tmpMin != tmpMax) {
+                        if (rank == 0) {
+                                WARN("inconsistent file size by different tasks");
```

Comment:
Nice check, albeit it costs a little performance.
As it assumes to be a collective operation that may lead to unexpected behavior (in terms of AIORI semantics) if not all processes invoke the same function. I'm not 100% sure if we should generally allow that but also not so worried as IO500 runs...